### DISCUSSION TITLE
Examples of Opentelemetry-cpp to support the Context Propagation via HTTP header

### DISCUSSION THREAD

**Discussion by r03942139** [open-telemetry/opentelemetry-cpp#929] [Q&A]
My goal is to implement a mechanism like money tracing [1] in the opentelemetry with Jaeger, they bypass trace context in the HTTP to realize distributed tracing.
And I found that the Opentelemetry core concepts [2] support such features, and there are some related implementations [3] related.
However, I found there are no examples about how to implement it.
Is it possible that I could have more information about this used cases?
Thanks.

References:
[1] Money Tracing: The process of origin will propagate that Trace Context across system boundaries by passing the Trace Context in an HTTP Header
https://github.com/Comcast/money/wiki#how-do-we-make-distributed-tracing-work
[2] OpenTelemetry Core Concepts: Context Propagation - Propagation via HTTP headers
https://opentelemetry.lightstep.com/core-concepts/context-propagation/
[3] Some implementations related to Context Propagation in Opentelemetry-cpp
https://github.com/open-telemetry/opentelemetry-cpp/issues/719

**Comment by lalitb** [comment]
@r03942139  - you can find an example for HTTP context propagation here: https://github.com/open-telemetry/opentelemetry-cpp/tree/main/examples/http

**Reply by r03942139** [reply]
@lalitb 
Can this example be expanded to support two exporters: OStreamSpanExporter, JaegerExporter at the same time.
The problem I hit is only the traces with error can be deployed, so I hope to debug through OStream.
Thanks.

The following is what I add to the example
https://github.com/open-telemetry/opentelemetry-cpp/blob/main/examples/http/tracer_common.h
----------------------- Start -----------------------
**#include "opentelemetry/trace/propagation/jaeger.h"**
**namespace trace    = opentelemetry::trace; 
namespace jaeger   = opentelemetry::exporter::jaeger; 
opentelemetry::exporter::jaeger::JaegerExporterOptions opts;**         

**auto exporter  = std::unique_ptr<sdktrace::SpanExporter>(new jaeger::JaegerExporter(opts)); // Change from OStream to Jaeger**

// Change from TextMapPropagator to JaegerPropagator
**opentelemetry::context::propagation::GlobalTextMapPropagator::SetGlobalPropagator(
      nostd::shared_ptr<opentelemetry::context::propagation::TextMapPropagator>(
          new opentelemetry::trace::propagation::JaegerPropagator()));**
----------------------- End -----------------------

The other parts remained the same.

**Reply by lalitb** [reply]
@r03942139  - you can create two processors, each one of them connected with a separate exporter. And add these two processors to a single TracerProvider. Please refer to the example below how it is done:

https://github.com/open-telemetry/opentelemetry-cpp/blob/v1.0.0-rc3/examples/multi_processor/main.cc

**Reply by r03942139** [reply]
@lalitb 
Thanks for the guidance, I used the following flow to implement context propagation with multi-processor.
The issue is that the traces will only be deployed to Jaeger, when the request is failed. (I assign to a wrong port)
When the request is successful (right port), the Jager cannot receive traces.

My question is: Do I need to add the **JaegerPropagator** as our **global propagator** either?
Thanks.

------- My flow of implementations -------
auto exporter1 = std::unique_ptr<sdktrace::SpanExporter>(
      new opentelemetry::exporter::trace::OStreamSpanExporter);
  auto processor1 = std::unique_ptr<sdktrace::SpanProcessor>(
      new sdktrace::SimpleSpanProcessor(std::move(exporter1)));

  auto exporter2  = std::unique_ptr<sdktrace::SpanExporter>(
      new jaeger::JaegerExporter(opts));
  auto processor2 = std::unique_ptr<sdktrace::SpanProcessor>(
      new sdktrace::SimpleSpanProcessor(std::move(exporter2)));

std::vector<std::unique_ptr<sdktrace::SpanProcessor>> processors;
  processors.push_back(std::move(processor1));
  processors.push_back(std::move(processor2));

  // Default is an always-on sampler.
  auto context  = std::make_shared<sdktrace::TracerContext>(std::move(processors));
  auto provider = nostd::shared_ptr<opentelemetry::trace::TracerProvider>(new sdktrace::TracerProvider(context));

  // Set the global trace provider
  opentelemetry::trace::Provider::SetTracerProvider(provider);

  // set global propagator
  opentelemetry::context::propagation::GlobalTextMapPropagator::SetGlobalPropagator(
      nostd::shared_ptr<opentelemetry::context::propagation::TextMapPropagator>(
          new opentelemetry::trace::propagation::HttpTraceContext()));
  // opentelemetry::context::propagation::GlobalTextMapPropagator::SetGlobalPropagator(
  //     nostd::shared_ptr<opentelemetry::context::propagation::TextMapPropagator>(
  //         new opentelemetry::trace::propagation::JaegerPropagator()));

-------------------------------------------------

However, the Jaeger UI only receive traces when 

------ Error will send traces ------
bash-5.0$ ./opentelemetry-cpp-1.0.0-rc3/build/examples/http/http_client 1010
{
  name          : /helloworld
  trace_id      : fdfa9f0395d8c4ee18e9808998d435a0
  span_id       : 17f2211fe7b5fe82
.........
  status        : **Error**

}
Sueccess request will not send traces to Jaeger
bash-5.0$ ./opentelemetry-cpp-1.0.0-rc3/build/examples/http/http_client 
{
  name          : /helloworld
  trace_id      : d7165329e6955168d8679f00f23d036b
  span_id       : 3610b691563a6524
  tracestate    : 
  parent_span_id: 0000000000000000
  start         : 1628110285581511155
  duration      : 1445337
  description   : 
  span kind     : Client
  status        : Unset
  attributes    : 
	http.header.Date: Wed, 04 Aug 2021 20:51:25 GMT
	http.header.Content-Type: text/plain
	http.header.Content-Length: 0
	http.header.Host: localhost
	http.scheme: http
	http.url: http://localhost:8800/helloworld
	http.method: GET
	http.header.Connection: keep-alive
	http.status_code: 200
  events        : 
  links         : 
}

**Reply by lalitb** [reply]
@r03942139 - Sorry, the issue is not very clear to me. If you are still facing the problem, can you please elaborate it a bit.
